src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py

```diff
@@ -26,195 +26,200 @@
-    cached = [[], []]
-    with torch.no_grad():
-        for img, t in dataloader:
-            try:
-                _ = model(img.to(device))
-            except StopFwdException:
-                pass
-            if store_inp:
-                if keep_gpu:
-                    cached[0].append(data_saver.input_store.detach())
-                else:
-                    cached[0].append(data_saver.input_store.detach().cpu())
-            if store_out:
-                if keep_gpu:
-                    cached[1].append(data_saver.output_store.detach())
-                else:
-                    cached[1].append(data_saver.output_store.detach().cpu())
-    if store_inp:
-        cached[0] = torch.cat([x for x in cached[0]])
-    if store_out:
-        cached[1] = torch.cat([x for x in cached[1]])
-    handle.remove()
-    if disable_quant:
-        disable_quant_class.enable_act_quantization(model, False)
-        disable_quant_class.enable_param_quantization(model, False)
-        restore_return_quant_tensor(model, return_quant_tensor_state)
-    return cached
+    img, _ = inputs
+    img = send_to_device(img, device)
+    model(img)
+
+
+def cnn_block_forward(block: nn.Module, inputs: Any) -> torch.Tensor:
+    device = next(block.parameters()).device
+    inputs = send_to_device(inputs, device)
+    return block(inputs)
+
+
+def apply_learned_round(
+    model: nn.Module,
+    calibration_loader: DataLoader,
+    iters: int = 1000,
+    learned_round: str = "hard_sigmoid_round",
+    learned_round_loss: str = "regularised_mse",
+    optimizer: str = "adam",
+    lr_scheduler: Optional[str] = None,
+    optimizer_lr: float = 1e-3,
+    batch_size: int = 1,
+    use_best_model: bool = False,
+    use_amp: bool = True,
+    amp_dtype: torch.dtype = torch.float16,
+    loss_scaling_factor: float = 1.,
+    learned_round_loss_kwargs: Optional[Dict] = None,
+    optimizer_kwargs: Optional[Dict] = None,
+    lr_scheduler_kwargs: Optional[Dict] = None,
+    learned_round_mode: str = "layerwise",
+) -> None:
+    if learned_round not in LEARNED_ROUND_MAP:
+        raise ValueError(f"Learned round method {learned_round} is not available.")
+    learned_round = LearnedRound(learned_round_impl_type=LEARNED_ROUND_MAP[learned_round])
+
+    if learned_round_loss not in LEARNED_ROUND_LOSS_MAP:
+        raise ValueError(f"Learned round loss {learned_round_loss} is not available.")
+    learned_round_loss_class = LEARNED_ROUND_LOSS_MAP[learned_round_loss]
+
+    if optimizer not in OPTIMIZER_MAP:
+        raise ValueError(f"Optimizer {optimizer} is not available.")
+    optimizer_class = OPTIMIZER_MAP[optimizer]
+
+    if lr_scheduler is not None and lr_scheduler not in LR_SCHEDULER_MAP:
+        raise ValueError(f"Learning rate scheduler {lr_scheduler} is not available.")
+    lr_scheduler_class = None if lr_scheduler is None else LR_SCHEDULER_MAP[lr_scheduler]
+
+    optimizer_classes = {"adam": torch.optim.Adam, "sign_sgd": SignSGD}
```

Comment:
Any particular reason why Adam? In other tests I've played a bit with SGD + momentum

Reply:
Maybe we can have a string that is used to fetch whatever optimizer the user wants, as lonog as it's an official pytorch one. i.e. `getattr(torch.optim, optimizer)`
The drawback of this is how to handle the various kwargs for each optimizer